File: app/src/main/java/com/mewo/reader/ui/search/SearchScreen.kt
package com.mewo.reader.ui.search

import android.content.Intent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.mewo.reader.data.BookRecord
import com.mewo.reader.data.FeedPost
import com.mewo.reader.ui.components.Avatar
import com.mewo.reader.ui.components.MewoAppBar
import com.mewo.reader.ui.reader.TimelinePost
import java.io.File

@Composable
fun SearchScreen(
    viewModel: SearchViewModel,
    onOpenBook: (String) -> Unit,
    onOpenPost: (bookId: String, postId: String) -> Unit,
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val focus = remember { FocusRequester() }
    LaunchedEffect(Unit) { focus.requestFocus() }

    Column(Modifier.fillMaxSize()) {
        MewoAppBar(
            content = {
                BasicTextField(
                    value = state.query,
                    onValueChange = viewModel::setQuery,
                    singleLine = true,
                    textStyle = MaterialTheme.typography.bodyLarge.merge(
                        TextStyle(color = MaterialTheme.colorScheme.onBackground),
                    ),
                    cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
                    modifier = Modifier
                        .weight(1f)
                        .padding(horizontal = 16.dp)
                        .focusRequester(focus),
                    decorationBox = { inner ->
                        if (state.query.isEmpty()) {
                            Text(
                                "Search books and posts",
                                style = MaterialTheme.typography.bodyLarge,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        inner()
                    },
                )
            },
        )
        HorizontalDivider(color = MaterialTheme.colorScheme.outline, thickness = 0.6.dp)

        when {
            state.query.isBlank() -> {
                Hint("Type a title, author, or a line you remember.")
            }
            state.result.books.isEmpty() && state.result.posts.isEmpty() -> {
                Hint("No matches for \"${state.query}\".")
            }
            else -> {
                LazyColumn(Modifier.fillMaxSize()) {
                    if (state.result.books.isNotEmpty()) {
                        item { SectionLabel("Books") }
                        items(state.result.books, key = { "book-${it.id}" }) { book ->
                            BookHit(
                                book = book,
                                cover = File(context.filesDir, "books/${book.id}/cover.jpg"),
                                onOpen = { onOpenBook(book.id) },
                            )
                            HorizontalDivider(
                                color = MaterialTheme.colorScheme.outline,
                                thickness = 0.6.dp,
                            )
                        }
                    }
                    if (state.result.posts.isNotEmpty()) {
                        item { SectionLabel("Posts") }
                        items(state.result.posts, key = { "post-${it.book.id}-${it.post.id}" }) { hit ->
                            TimelinePost(
                                post = hit.post,
                                author = hit.book.author,
                                handle = hit.book.handle,
                                cover = File(context.filesDir, "books/${hit.book.id}/cover.jpg"),
                                liked = state.likes[hit.book.id]?.contains(hit.post.id) == true,
                                onLike = { viewModel.toggleLike(hit.book.id, hit.post.id) },
                                onRepost = {
                                    shareText(context, quote(hit.book.author, hit.book.title, hit.post))
                                },
                                onShare = { shareText(context, hit.post.text) },
                                onOpen = { onOpenPost(hit.book.id, hit.post.id) },
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text = text,
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(horizontal = 16.dp, vertical = 10.dp),
    )
}

@Composable
private fun Hint(text: String) {
    Text(
        text = text,
        style = MaterialTheme.typography.bodyLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(horizontal = 16.dp, vertical = 24.dp),
    )
}

@Composable
private fun BookHit(
    book: BookRecord,
    cover: File,
    onOpen: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(role = Role.Button, onClick = onOpen)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Avatar(name = book.author, coverFile = cover, size = 40.dp)
        Spacer(Modifier.width(12.dp))
        Column {
            Text(book.title, style = MaterialTheme.typography.titleSmall)
            Spacer(Modifier.height(2.dp))
            Text(
                text = "@${book.handle}",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

private fun quote(author: String?, title: String?, post: FeedPost): String {
    val by = listOfNotNull(author, title).joinToString(", ")
    return if (by.isBlank()) post.text else "${post.text}\n\n— $by"
}

private fun shareText(context: android.content.Context, text: String) {
    val intent = Intent(Intent.ACTION_SEND).apply {
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, text)
    }
    context.startActivity(Intent.createChooser(intent, null))
}

File: ios/Mewo/UI/Search/SearchScreen.swift
import SwiftUI

@Observable
@MainActor
final class SearchViewModel {
    var query = ""
    var result = SearchResult()
    var likes: [String: Set<String>] = [:]
    var comments: [String: CommentIndex] = [:]
    private let env: AppEnvironment

    init(env: AppEnvironment) { self.env = env }

    func setQuery(_ query: String) {
        self.query = query
        Task {
            do {
                let hits = try await env.library.search(query: query)
                let bookIds = Array(Set(hits.posts.map(\.book.id)))
                var likes: [String: Set<String>] = [:]
                var comments: [String: CommentIndex] = [:]
                for id in bookIds {
                    likes[id] = try await env.library.likes(id: id)
                    comments[id] = (try? await env.library.commentIndex(id: id)) ?? CommentIndex()
                }
                if self.query == query {
                    self.result = hits
                    self.likes.merge(likes) { _, new in new }
                    self.comments.merge(comments) { _, new in new }
                }
            } catch {}
        }
    }

    func toggleLike(bookId: String, postId: String) {
        Task {
            if let next = try? await env.library.toggleLike(id: bookId, postId: postId) {
                likes[bookId] = next
            }
        }
    }

    func onCommentsChanged(bookId: String, postId: String, comments: [PostComment]) {
        let current = self.comments[bookId] ?? CommentIndex()
        self.comments[bookId] = current.afterThread(postId: postId, comments: comments)
    }
}

struct SearchScreen: View {
    @Environment(\.mewo) private var mewo
    @Environment(\.app) private var app
    @Bindable var chrome: HideOnScrollState
    var onOpenBook: (String) -> Void
    var onOpenPost: (String, String) -> Void
    var onOpenAccount: () -> Void
    var onOpenProfile: (String, String) -> Void
    @State private var model: SearchViewModel
    @State private var commentTarget: CommentTarget?
    @State private var shareText: String?
    @State private var lastOffset: CGFloat = 0
    @FocusState private var focused: Bool

    init(env: AppEnvironment, chrome: HideOnScrollState, onOpenBook: @escaping (String) -> Void, onOpenPost: @escaping (String, String) -> Void, onOpenAccount: @escaping () -> Void, onOpenProfile: @escaping (String, String) -> Void) {
        self.chrome = chrome
        self.onOpenBook = onOpenBook
        self.onOpenPost = onOpenPost
        self.onOpenAccount = onOpenAccount
        self.onOpenProfile = onOpenProfile
        _model = State(initialValue: SearchViewModel(env: env))
    }

    var body: some View {
        VStack(spacing: 0) {
            MewoAppBar(leading: { ProfileButton(onClick: onOpenAccount) }) {
                TextField("Search books and posts", text: Binding(
                    get: { model.query },
                    set: { model.setQuery($0) }
                ))
                .font(MewoFont.regular(17))
                .foregroundStyle(mewo.ink)
                .focused($focused)
                .padding(.trailing, 16)
                .accessibilityIdentifier("search-field")
            }
            results
        }
        .background(mewo.ground)
        .onAppear { focused = true }
        .onChange(of: model.query) { _, q in
            chrome.showWhenIdle(q.isEmpty || (model.result.books.isEmpty && model.result.posts.isEmpty))
        }
        .sheet(item: $commentTarget) { target in
            CommentSheet(
                target: target,
                publicNotes: app.backend.kind == .hosted,
                onThreadChanged: { postId, comments in
                    model.onCommentsChanged(bookId: target.book.id, postId: postId, comments: comments)
                },
                onDismiss: { commentTarget = nil },
                onOpenProfile: { handle, name in
                    commentTarget = nil
                    onOpenProfile(handle, name)
                }
            )
        }
        .sheet(isPresented: Binding(get: { shareText != nil }, set: { if !$0 { shareText = nil } })) {
            if let shareText { ShareSheet(text: shareText) }
        }
    }

    @ViewBuilder
    private var results: some View {
        if model.query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            hint("Type a title, author, or a line you remember.")
        } else if model.result.books.isEmpty && model.result.posts.isEmpty {
            hint("No matches for \"\(model.query)\".")
        } else {
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 0) {
                    GeometryReader { geo in
                        Color.clear.preference(key: ScrollOffsetKey.self, value: -geo.frame(in: .named("search")).minY)
                    }
                    .frame(height: 0)
                    if !model.result.books.isEmpty {
                        section("Books")
                        ForEach(model.result.books) { book in
                            Button { onOpenBook(book.id) } label: {
                                HStack(spacing: 12) {
                                    Avatar(name: book.author, cover: app.library.coverFile(id: book.id), size: 40)
                                    VStack(alignment: .leading, spacing: 2) {
                                        Text(book.title).font(MewoFont.bold(15)).foregroundStyle(mewo.ink)
                                        Text("@\(book.handle)").font(MewoFont.regular(13)).foregroundStyle(mewo.mute)
                                    }
                                    Spacer()
                                }
                                .padding(.horizontal, 16)
                                .padding(.vertical, 12)
                            }
                            .buttonStyle(.plain)
                            Hairline()
                        }
                    }
                    if !model.result.posts.isEmpty {
                        section("Posts")
                        ForEach(model.result.posts) { hit in
                            TimelinePost(
                                post: hit.post,
                                author: hit.book.author,
                                handle: hit.book.handle,
                                cover: app.library.coverFile(id: hit.book.id),
                                liked: model.likes[hit.book.id]?.contains(hit.post.id) == true,
                                commentCount: model.comments[hit.book.id]?.count(hit.post.id) ?? 0,
                                commented: model.comments[hit.book.id]?.commented(hit.post.id) == true,
                                onComment: { commentTarget = CommentTarget(book: hit.book, post: hit.post) },
                                onLike: { model.toggleLike(bookId: hit.book.id, postId: hit.post.id) },
                                onRepost: { shareText = quoteLine(author: hit.book.author, title: hit.book.title, text: hit.post.text) },
                                onShare: { shareText = hit.post.text },
                                onOpen: { onOpenPost(hit.book.id, hit.post.id) }
                            )
                        }
                    }
                }
            }
            .coordinateSpace(name: "search")
            .onPreferenceChange(ScrollOffsetKey.self) { offset in
                chrome.ingest(delta: offset - lastOffset)
                lastOffset = offset
                chrome.bindTop(offset)
            }
        }
    }

    private func section(_ text: String) -> some View {
        Text(text)
            .font(MewoFont.regular(13))
            .foregroundStyle(mewo.mute)
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
    }

    private func hint(_ text: String) -> some View {
        Text(text)
            .font(MewoFont.regular(17))
            .foregroundStyle(mewo.mute)
            .padding(.horizontal, 16)
            .padding(.vertical, 24)
            .frame(maxWidth: .infinity, alignment: .leading)
    }
}
